### PR TITLE
hw

### PR DESCRIPTION
**Pull request by Sp0ozy**

### CODE REVIEW

File: exercises/any_odds_and_evens/main_test.go [hunk 1]
```diff
@@ -0,0 +1,23 @@
+package main
+
+import "testing"
+
+func TestAnyIsOdd(t *testing.T) {
+	for _, tc := range []struct {
+		s    []int
+		want bool
+	}{
+		{[]int{}, false},
```
Comment (zayac) [inline exercises/any_odds_and_evens/main_test.go:10]: What about tests containing zeros/negative even numbers?

File: exercises/any_odds_and_evens/main_test.go [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+		{[]int{2}, false},
+		{[]int{3}, true},
+		{[]int{2, 4, 8}, false},
+		{[]int{2, 1, 8}, true},
+		{[]int{3, 7, 9}, true},
+		{[]int{-3, 1, 8}, true},
+	} {
+		got := AnyIsOdd(tc.s)
```
Comment (zayac) [inline exercises/any_odds_and_evens/main_test.go]: nit: consider writing these lines as one:
```go
if got := AnyIsEven(ts.s); got != tc.want {
  ...
}
```

Same in other programs.

File: exercises/only_odds_and_evens/main_test.go [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+import "testing"
+
+func TestAreOddsOnly(t *testing.T) {
+	for _, tc := range []struct {
+		s    []int
+		want bool
+	}{
+		{[]int{}, true},
+		{[]int{0}, false},
```
Comment (zayac) [inline exercises/only_odds_and_evens/main_test.go:11]: nit: you could add a test containing a single negative number.

File: exercises/filter_odds_and_evens/main.go [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+}
+
+func FilterOdds(s []int) (a []int) {
+	for _, v := range s {
+		if v%2 == 1 {
+			a = append(a, v)
+		}
+		if v%2 == -1 {
+			a = append(a, v)
+		}
```
Comment (zayac) [inline exercises/filter_odds_and_evens/main.go]: You can combine these two `if`s in a single `if`: `if v%2 != 0`

File: exercises/filter_odds_and_evens/main_test.go [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+	for _, tc := range []struct {
+		s    []int
+		want []int
+	}{
+		{[]int{}, []int{}},
+		{[]int{0}, []int{0}},
+		{[]int{0, 1}, []int{0}},
+		{[]int{-1, -2}, []int{-2}},
+		{[]int{3, 7, 4, 9}, []int{4}},
+		{[]int{4, 6, 8, 1, 3, 91}, []int{4, 6, 8}},
```
Comment (zayac) [inline exercises/filter_odds_and_evens/main_test.go:26]: nit: consider some negatives here as well

File: exercises/rect/main_test.go [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+func equal(a, b [][]int) bool {
+	return reflect.DeepEqual(a, b)
+}
+func TestRect(t *testing.T) {
+	for _, tc := range []struct {
+		x, y int
+		want [][]int
+	}{
+		{0, 0, [][]int{}},
```
Comment (zayac) [inline exercises/rect/main_test.go:16]: nit: consider adding `{0, X}` and `{X, 0}` (where `X > 0`). Also negatives

File: exercises/spiral/main_test.go [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+
+func equal(a, b [][]int) bool {
+	return reflect.DeepEqual(a, b)
+}
+func TestSpiral(t *testing.T) {
+	for _, tc := range []struct {
+		x, y int
+		want [][]int
+	}{
+		{0, 0, [][]int{}},
```
Comment (zayac) [inline exercises/spiral/main_test.go:16]: nit: consider adding tests for (0, x), (x, 0) and negative inputs

File: exercises/spiral/main.go [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+	for i := range s {
+		s[i] = make([]int, x)
+	}
+
+	l := 1
+
+	a := 0
+	b := 0
+	xmin := 0
+	ymin := 0
```
Comment (zayac) [inline exercises/spiral/main.go]: Use a shorter initialization instead:

```
var a, b, xmin, ymin int
```

File: exercises/spiral/main.go [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+			b--
+		}
+		b++
+		xmin++
+		a++
+
+	}
+	if x == 1 && y == 1 {
+		s[0] = []int{1}
+	}
```
Comment (zayac) [inline exercises/spiral/main.go:62]: Why do you need this special condition?

Comment (Sp0ozy) [reply exercises/spiral/main.go:62]: If the height and width are 1, the function fills the rectangle with 0.

File: exercises/spiral/main_test.go [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+		}},
+		{3, 7, [][]int{
+			{1, 2, 3},
+			{16, 17, 4},
+			{15, 24, 5},
+			{14, 23, 6},
+			{13, 22, 7},
+			{12, 21, 8},
+			{11, 10, 9},
+		}},
```
Comment (zayac) [inline exercises/spiral/main_test.go:38]: This is not a correct output. It should be

```
		{3, 7, [][]int{
			{1, 2, 3},
			{16, 17, 4},
			{15, 18, 5},
			{14, 19, 6},
			{13, 20, 7},
			{12, 21, 8},
			{11, 10, 9},
		}},
```
